### ISSUE TITLE
Correct way to incorporate train and validation datasets in SFTTrainer

### ISSUE THREAD

**Issue by karths8** [huggingface/trl#603]
I used the following [script](https://gist.github.com/younesbelkada/9f7f75c94bdc1981c8ca5cc937d4a4da) that uses SFTTrainer to train Llama-2 on my own dataset using QLoRA. I am quite stuck on how to format the validation dataset in this case. From this [link](https://www.philschmid.de/instruction-tune-llama-2) it seems apparent that the train_data for the SFTTrainer would have a format similar to this:

```
### Instruction:
{sample['instruction']}

### Input:
{sample['input']}

### Response:
{sample['response']}
```

First of all, I was curious why this was the format to be passed into the Trainer. How does the SFTTrainer work behind the scenes to turn this into an input/output pair suitable for training?
Secondly, I wanted to know how I should format and pass in my validation dataset in the proper manner

Please let me know!

**Comment by lvwerra** [comment]
This is a common format for instruction tuning models. In principle it can be any function that takes a row of your dataset and returns a text field. This function is applied to the whole dataset then. Note that since this is language modeling, there are no input output pairs but the whole text is input and label at the same time. 

You should pass the validation set in the same format as the training dataset as they are treated the same under the hood.

Hope this helps!

**Comment by lvwerra** [comment]
This seems to be a different issue. Are you using quantization and LoRA? cc @younesbelkada could be similar to our issues with BF16 vs FP16.

**Comment by sunzx8** [comment]
> This seems to be a different issue. Are you using quantization and LoRA? cc @younesbelkada could be similar to our issues with BF16 vs FP16.

No, I do not use LoRA or quantization,this is my whole code
`# Based on younesbelkada/finetune_llama_v2.py
# Install the following libraries:
# pip install accelerate==0.21.0 peft==0.4.0 bitsandbytes==0.40.2 transformers==4.31.0 trl==0.4.7 scipy

from dataclasses import dataclass, field
from typing import Optional
import torch.distributed
import torch
from datasets import load_dataset
from transformers import (
    AutoModelForCausalLM,
    AutoTokenizer,
    BitsAndBytesConfig,
    HfArgumentParser,
    TrainingArguments,
    HfArgumentParser,
    Trainer,
    AutoModelForMultipleChoice
)

from trl import SFTTrainer
import wandb

@dataclass
class ScriptArguments:
    model_name: Optional[str] = field(
        default="/mnt/workspace/workgroup/hanxiao/llama2/Llama-2-7b-hf",
        metadata={"help": "The model that you want to train from the Hugging Face hub"},
    )
    dataset_name: Optional[str] = field(
        default="/mnt/workspace/workgroup/hanxiao/llama2/llama-recipes/ft_datasets/alpaca_data.json",
        metadata={"help": "The instruction dataset to use"},
    )
    new_model: Optional[str] = field(
        default="llama-science", metadata={"help": "Fine-tuned model name"}
    )
    merge_and_push: Optional[bool] = field(
        default=True, metadata={"help": "Merge and push weights after training"}
    )

    # QLoRA parameters
    lora_r: Optional[int] = field(
        default=64, metadata={"help": "LoRA attention dimension"}
    )
    lora_alpha: Optional[int] = field(
        default=16, metadata={"help": "Alpha parameter for LoRA scaling"}
    )
    lora_dropout: Optional[float] = field(
        default=0.1, metadata={"help": "Dropout probability for LoRA layers"}
    )

    # bitsandbytes parameters
    use_4bit: Optional[bool] = field(
        default=False, metadata={"help": "Activate 4-bit precision base model loading"}
    )
    bnb_4bit_compute_dtype: Optional[str] = field(
        default="float16", metadata={"help": "Compute dtype for 4-bit base models"}
    )
    bnb_4bit_quant_type: Optional[str] = field(
        default="nf4", metadata={"help": "Quantization type (fp4 or nf4)"}
    )
    use_nested_quant: Optional[bool] = field(
        default=False,
        metadata={
            "help": "Activate nested quantization for 4-bit base models (double quantization)"
        },
    )

    # TrainingArguments parameters
    output_dir: str = field(
        default="/mnt/workspace/workgroup/hanxiao/llama2/llama-science",
        metadata={
            "help": "Output directory where the model predictions and checkpoints will be stored"
        },
    )
    num_train_epochs: Optional[int] = field(
        default=3, metadata={"help": "Number of training epochs"}
    )
    fp16: Optional[bool] = field(
        default=False, metadata={"help": "Enable fp16 training"}
    )
    bf16: Optional[bool] = field(
        default=False, metadata={"help": "Enable bf16 training"}
    )
    tf32: Optional[bool] = field(
        default=True,
        metadata={"help": "Enable the TF32 mode (available in Ampere and newer GPUs)"},
    )
    per_device_train_batch_size: Optional[int] = field(
        default=1, metadata={"help": "Batch size per GPU for training"}
    )
    per_device_eval_batch_size: Optional[int] = field(
        default=12, metadata={"help": "Batch size per GPU for evaluation"}
    )
    gradient_accumulation_steps: Optional[int] = field(
        default=1,
        metadata={"help": "Number of update steps to accumulate the gradients for"},
    )
    gradient_checkpointing: Optional[bool] = field(
        default=True, metadata={"help": "Enable gradient checkpointing"}
    )
    max_grad_norm: Optional[float] = field(
        default=0.3, metadata={"help": "Maximum gradient normal (gradient clipping)"}
    )
    learning_rate: Optional[float] = field(
        default=1e-5, metadata={"help": "Initial learning rate (AdamW optimizer)"}
    )
    weight_decay: Optional[int] = field(
        default=0.001,
        metadata={
            "help": "Weight decay to apply to all layers except bias/LayerNorm weights"
        },
    )
    optim: Optional[str] = field(
        default="adamw_hf", metadata={"help": "Optimizer to use"}
    )
    lr_scheduler_type: str = field(
        default="cosine",
        metadata={"help": "Learning rate schedule"},
    )
    max_steps: int = field(
        default=-1,
        metadata={"help": "Number of training steps (overrides num_train_epochs)"},
    )
    warmup_ratio: float = field(
        default=0.01,
        metadata={
            "help": "Ratio of steps for a linear warmup (from 0 to learning rate)"
        },
    )
    group_by_length: bool = field(
        default=True,
        metadata={
            "help": "Group sequences into batches with same length. Saves memory and speeds up training considerably"
        },
    )
    save_steps: float = field(
        default=0, metadata={"help": "Save checkpoint every X updates steps"}
    )
    logging_steps: int = field(
        default=1, metadata={"help": "Log every X updates steps"}
    )
    resume_from_checkpoint: bool = field(
        default=False, 
        metadata={"help": "Allows to resume training from the latest checkpoint in output_dir"}
    )

    # SFT parameters
    max_seq_length: Optional[int] = field(
        default=2048, metadata={"help": "Maximum sequence length to use"}
    )
    packing: Optional[bool] = field(
        default=False,
        metadata={
            "help": "Pack multiple short examples in the same input sequence to increase efficiency"
        },
    )


parser = HfArgumentParser(ScriptArguments)
script_args = parser.parse_args_into_dataclasses()[0]
device_map = "auto"

# Load dataset (you can process it here)
dataset = load_dataset('json',data_files=script_args.dataset_name, split="train")

# Load tokenizer and model with QLoRA configuration
compute_dtype = getattr(torch, script_args.bnb_4bit_compute_dtype)



# Check GPU compatibility with bfloat16
if compute_dtype == torch.float16 and script_args.use_4bit:
    major, _ = torch.cuda.get_device_capability()
    if major >= 8:
        print("=" * 80)
        print("Your GPU supports bfloat16: accelerate training with --bf16")
        print("=" * 80)

# Load base model
model = AutoModelForMultipleChoice.from_pretrained(
    script_args.model_name,
    torch_dtype=torch.float16,
    device_map=device_map,
)
model.config.use_cache = False
model.config.pretraining_tp = 1

# Load LLaMA tokenizer
tokenizer = AutoTokenizer.from_pretrained(
    script_args.model_name, trust_remote_code=True
)
tokenizer.pad_token = tokenizer.eos_token
tokenizer.padding_side = "right"  # Fix weird overflow issue with fp16 training

# # Load LoRA configuration
# peft_config = LoraConfig(
#     lora_alpha=script_args.lora_alpha,
#     lora_dropout=script_args.lora_dropout,
#     r=script_args.lora_r,
#     bias="none",
#     task_type="CAUSAL_LM",
# )

# Set training parameters
training_arguments = TrainingArguments(
    output_dir=script_args.output_dir,
    num_train_epochs=script_args.num_train_epochs,
    per_device_train_batch_size=script_args.per_device_train_batch_size,
    gradient_accumulation_steps=script_args.gradient_accumulation_steps,
    optim=script_args.optim,
    save_steps=script_args.save_steps,
    logging_steps=script_args.logging_steps,
    learning_rate=script_args.learning_rate,
    weight_decay=script_args.weight_decay,
    fp16=script_args.fp16,
    bf16=script_args.bf16,
    max_grad_norm=script_args.max_grad_norm,
    max_steps=script_args.max_steps,
    warmup_ratio=script_args.warmup_ratio,
    group_by_length=script_args.group_by_length,
    lr_scheduler_type=script_args.lr_scheduler_type,
    report_to="all",
)
def train_tokenize_function(examples, tokenizer):
    prompt_input= PROMPT_DICT["prompt_input"]
    sources = [
            prompt_input.format_map(dict(Instruction=Instruction))\
            for Instruction in zip(examples['Instruction']) 
        ]
    targets = [f"{output}{tokenizer.eos_token}" for output in examples['Response']]
    data_dict = preprocess(sources, targets, tokenizer)
    return data_dict

def preprocess(
    sources: Sequence[str],
    targets: Sequence[str],
    tokenizer: transformers.PreTrainedTokenizer,
) -> Dict:
    """Preprocess the data by tokenizing."""
    examples = [s + t for s, t in zip(sources, targets)]
    examples_tokenized, sources_tokenized = [_tokenize_fn(strings, tokenizer) for strings in (examples, sources)]
    input_ids = examples_tokenized["input_ids"]
    labels = copy.deepcopy(input_ids)
    for label, source_len in zip(labels, sources_tokenized["input_ids_lens"]):
        label[:source_len] = IGNORE_INDEX
    return dict(input_ids=input_ids, labels=labels)

def formatting_prompts_func(example):
    a_list=[]
    for  i in range(len(example['input'])):
        text = (
            f"### Instruction: {example['instruction'][i]}\n" +
            f"### Input: {example['input'][i]}\n" +
            f"### Response: {example['output'][i]}"
        )
        # print(text)
        a_list.append(text)
    return a_list

# Set supervised fine-tuning parameters
trainer = SFTTrainer(
    model=model,
    train_dataset=dataset,
    formatting_func=formatting_prompts_func,
    max_seq_length=script_args.max_seq_length,
    tokenizer=tokenizer,
    args=training_arguments,
)
wandb.init(
      entity='llm-text2sql',
      # Set the project where this run will be logged
      project="hanxiao", 
      # We pass a run name (otherwise it’ll be randomly assigned, like sunshine-lollypop-10)
)
# Train model
trainer.train(resume_from_checkpoint=script_args.resume_from_checkpoint)

# Save trained model
trainer.model.save_pretrained(script_args.new_model)`
I have no idea what's going on with my code traning but the loss keep 0
<img width="526" alt="image" src="https://github.com/huggingface/trl/assets/136688632/1f311897-f53f-4734-92d3-5fc9ba94cd31">

before I always used preprocess like
`def preprocess(
    sources: Sequence[str],
    targets: Sequence[str],
    tokenizer: transformers.PreTrainedTokenizer,
) -> Dict:
    """Preprocess the data by tokenizing."""
    examples = [s + t for s, t in zip(sources, targets)]
    examples_tokenized, sources_tokenized = [_tokenize_fn(strings, tokenizer) for strings in (examples, sources)]
    input_ids = examples_tokenized["input_ids"]
    labels = copy.deepcopy(input_ids)
    for label, source_len in zip(labels, sources_tokenized["input_ids_lens"]):
        label[:source_len] = IGNORE_INDEX
    return dict(input_ids=input_ids, labels=labels)
`

**Comment by RManLuo** [comment]
Similar question, it appears to occur exclusively when utilizing FSDP. I observe that I experience the correct loss when I make the switch to using DeepSpeed.

**Comment by ZSGS-Vipin** [comment]
`from trl import SFTTrainer

trainer = SFTTrainer(
    model = model,
    tokenizer=tokenizer,
    train_dataset=train_dataset,
    eval_dataset=validation_dataset,
    args = training_args,
    max_seq_length=2048,
    packing=False,
    # num_of_sequences=1024,
    formatting_func=formatting_prompts_func,
    data_collator=data_collator,
    compute_metrics=compute_metrics,
    preprocess_logits_for_metrics = preprocess_logits_for_metrics
)

trainer.train()`

Is there any way to save this mapped train and test set, so that there will be no waste of time to map it again and again
